### DISCUSSION TITLE
Cannot build the tree-sitter-python grammar using Rust x86_64-unknown-linux-musl target

### DISCUSSION THREAD

**Discussion by Luni-4** [tree-sitter/tree-sitter#794] [Q&A] [answered]
I tried to build the `tree-sitter-python` project described [here](https://github.com/tree-sitter/tree-sitter-python/tree/master/bindings/rust) using the Rust `x86_64-unknown-linux-musl` target but it seems the compiler can't be found. Perhaps adding the `.compiler("gcc")` option [here](https://github.com/tree-sitter/tree-sitter-python/blob/master/bindings/rust/build.rs#L22)?

Thanks in advance for your help! :)

**Comment by maxbrunsfeld** [comment, marked as answer]
Do you have `gcc` and `g++` on your `PATH`? It should find them if so.

**Reply by Luni-4** [reply]
Yep, I can build it smoothly using `cargo build`. 

Doesn't it happen to you when you run: `cargo build --target x86_64-unknown-linux-musl` ?

**Reply by maxbrunsfeld** [reply]
Can you post the error message?

**Reply by Luni-4** [reply]
Here it is:

```
Caused by:
  process didn't exit successfully: `/my/home/try-tree-sitter/target/debug/build/tree-sitter-python-43fbe13b072bceda/build-script-build` (exit code: 1)
  --- stdout
  cargo:rerun-if-changed=src/parser.c
  TARGET = Some("x86_64-unknown-linux-musl")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-unknown-linux-gnu")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "musl-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "src" "-Wall" "-Wextra" "-Wno-unused-parameter" "-Wno-unused-but-set-variable" "-Wno-trigraphs" "-o" "/my/home/try-tree-sitter/target/x86_64-unknown-linux-musl/debug/build/tree-sitter-python-bb904181a1929c59/out/src/parser.o" "-c" "src/parser.c"
  exit code: 0
  AR_x86_64-unknown-linux-musl = None
  AR_x86_64_unknown_linux_musl = None
  TARGET_AR = None
  AR = None
  running: "ar" "crs" "/my/home/try-tree-sitter/target/x86_64-unknown-linux-musl/debug/build/tree-sitter-python-bb904181a1929c59/out/libparser.a" "/my/home/try-tree-sitter/target/x86_64-unknown-linux-musl/debug/build/tree-sitter-python-bb904181a1929c59/out/src/parser.o"
  exit code: 0
  cargo:rustc-link-lib=static=parser
  cargo:rustc-link-search=native=/my/home/try-tree-sitter/target/x86_64-unknown-linux-musl/debug/build/tree-sitter-python-bb904181a1929c59/out
  cargo:rerun-if-changed=src/scanner.cc
  TARGET = Some("x86_64-unknown-linux-musl")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-unknown-linux-gnu")
  CXX_x86_64-unknown-linux-musl = None
  CXX_x86_64_unknown_linux_musl = None
  TARGET_CXX = None
  CXX = None
  CROSS_COMPILE = None
  CXXFLAGS_x86_64-unknown-linux-musl = None
  CXXFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  CXX_x86_64-unknown-linux-musl = None
  CXX_x86_64_unknown_linux_musl = None
  TARGET_CXX = None
  CXX = None
  CROSS_COMPILE = None
  CXXFLAGS_x86_64-unknown-linux-musl = None
  CXXFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  CXX_x86_64-unknown-linux-musl = None
  CXX_x86_64_unknown_linux_musl = None
  TARGET_CXX = None
  CXX = None
  CROSS_COMPILE = None
  CXXFLAGS_x86_64-unknown-linux-musl = None
  CXXFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "musl-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "src" "-Wall" "-Wextra" "-o" "/my/home/try-tree-sitter/target/x86_64-unknown-linux-musl/debug/build/tree-sitter-python-bb904181a1929c59/out/src/scanner.o" "-c" "src/scanner.cc"
    --- stderr


  error occurred: Failed to find tool. Is `musl-g++` installed?
```

**Reply by maxbrunsfeld** [reply]
Oh, I guess you need an executable called musl-c++?

I’m not very familiar with musl, so I’m not sure why the ‘cc’ crate is trying to invoke that tool.

**Reply by Luni-4** [reply]
Yay! :smile: 

I built `tree-sitter-python` locally adding `compiler("gcc")` to the `build.rs` file, now it works perfectly! It didn't affect the `cargo build` command either. Do you want a PR?

**Reply by maxbrunsfeld** [reply]
Hmm, I would prefer that it use the default compiler instead of hard-coding gcc. What if you just set the `CC=gcc CXX=g++` environment variables when running `cargo build`?

**Reply by Luni-4** [reply]
It works the same! Thank you! Perhaps adding this to the README explaining how to build using `musl`?

**Reply by harrysarson** [reply]
So the issue here is the most straightforward way to get a musl compiler is `apt install musl-tools` (or `pacman -Syu musl` or whatever for redhat). These packages provide `musl-gcc` with a musl based `libc.a`. They do not provide `musl-g++` nor a musl based `libstdc++.a` (no idea why not). This explains why the `build.rs` script cannot find `musl-g++`: it simply does not exist.

The only way I know of to get a `musl-g++` executable is to compile it your self (not that hard if you use https://github.com/richfelker/musl-cross-make).

> What if you just set the `CC=gcc CXX=g++` environment variables when running `cargo build`?

This kind of works! It uses the system `gcc` and `g++` to compiler the tree-sitter c/c++ files into static libraries and then `musl-gcc` to link everything (including those static libraries) together into a static executable. The possible problem is that there may (unlikely I think) be an ABI mismatch between the system `gcc` (that uses one libc/libstdc++) and (`musl-gcc` that uses a different libc/libstdc++`). 

The [musl website](https://www.musl-libc.org/faq.html) has the following to say about compiling c++ programs.

> Be aware that, "out of the box", the wrapper only supports C applications, not C++. This is because the C++ libraries and headers are missing from the musl include/library path. The existing libstdc++ is actually compatible with musl in most cases and could be used by copying it into the musl library path, but the C++ header files are usually not compatible. One option may be rebuilding just libstdc++ against musl; however, if C++ support is needed, it's recommended just to build a native toolchain targetting musl.
